### PR TITLE
HoldToRevealButton component

### LINKED ISSUE #13260: Develop "Hold to reveal" component
**User Story**
Develop "Hold to reveal" component/control 

**Figma link:** 
https://www.figma.com/file/8z4Csc3wlOIkUz2OP1eTZW/Secure-UX-Improvements?node-id=44%3A1387)

**Task(s)**
Develop "Hold to reveal" component, according to the Figma mockup

**Acceptance criteria**
- Develop "Hold to reveal" component, according to the Figma mockup below. Please see design and functionality on the link:
https://codepen.io/antonela/full/eYEKRxL.

image

- The component will have two properties: Label, and Time in seconds.
- The component will have a callback mechanism to notify the parent component about the result (completion).

Notes: 
- We need to develop the unit tests for the component/control
- Consensys (Zoom meeting 10.02.2022) decided to develop the component in smaller size from the figma design

### PR DESCRIPTION
**Pull request by filipsekulic**
Fixes: https://github.com/MetaMask/metamask-extension/issues/13260

Screencast:

https://user-images.githubusercontent.com/92310504/155983424-8d453271-9484-4019-9f69-c9af366d3b21.mov

### CODE REVIEW

**Comment by mirjanaKukic** [comment]
Verified by QA

File: ui/components/app/hold-to-reveal-button/index.scss [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+  width: $circle-diameter;
+}
+
+%circle-shared {
+  fill: transparent;
+  stroke-width: $circle-stroke-width;
+}
+
+// Class styles
+.button-hold {
```
Comment (mcmire) [inline ui/components/app/hold-to-reveal-button/index.scss:59]: Can we use [BEM](https://en.bem.info/methodology/quick-start/#introduction) here? Remember that all of these styles are global and could therefore clash with other classes we add in the future. So all of these would become

``` sass
.hold-to-reveal-button {
  &__button-hold {
    /* ... */
  }
  
  &__absolute-fill {
    /* ... */
  }

  /* ... etc ... */
}
```

Also is there a reason why we are using `!important`?

File: ui/components/app/hold-to-reveal-button/index.scss [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+    opacity: 0;
+  }
+
+  to {
+    opacity: 1;
+  }
+}
+
+// Shared styles
+%absolute-fill {
```
Comment (mcmire) [inline ui/components/app/hold-to-reveal-button/index.scss:40]: Can we scope these mixins within a `.hold-to-reveal-button` class? That way they don't get created globally.

File: ui/components/app/hold-to-reveal-button/hold-to-reveal-button.js [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+   */
+  const resetAnimationStates = () => {
+    setIsUnlocking(false);
+    setHasTriggeredUnlock(false);
+  };
+
+  const renderPreCompleteContent = useCallback(() => {
+    return (
+      <Box
+        className={`absolute-fill ${isUnlocking ? 'invisible' : null} ${
```
Comment (mcmire) [inline ui/components/app/hold-to-reveal-button/hold-to-reveal-button.js:74]: [nit] You can use `classnames` to clean this up a bit if you like — should allow you to split this across multiple lines.

File: ui/components/app/hold-to-reveal-button/hold-to-reveal-button.js [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+   */
+  const preventPropogation = (e) => {
+    e.stopPropagation();
+  };
+
+  /**
+   * Event for mouse click down
+   */
+  const onMouseDown = () => {
+    isLongPressing.current = true;
```
Comment (mcmire) [inline ui/components/app/hold-to-reveal-button/hold-to-reveal-button.js:35]: Do you mind explaining how this works? All I'm seeing is that if you mouse down, this line gets run but I'm a bit confused as to how the other functions here end up getting called.

File: ui/components/app/hold-to-reveal-button/index.scss (hunk 1 above)
Comment (filipsekulic) [reply ui/components/app/hold-to-reveal-button/index.scss]: Fixed in: 2afe9c604ab6d1e79ede7a1d02693f1a34d96e3c
The reason for using the `!important` is 'overriding' some properties of the `Button` component that is being used as the wrapper component in this case.

File: ui/components/app/hold-to-reveal-button/index.scss (hunk 2 above)
Comment (filipsekulic) [reply ui/components/app/hold-to-reveal-button/index.scss]: Fixed in: 2afe9c604ab6d1e79ede7a1d02693f1a34d96e3c

File: ui/components/app/hold-to-reveal-button/hold-to-reveal-button.js (hunk 4 above)
Comment (filipsekulic) [reply ui/components/app/hold-to-reveal-button/hold-to-reveal-button.js:35]: You can check the ticket description (https://github.com/MetaMask/metamask-extension/issues/13260) and the conversation on Slack (https://consensys.slack.com/archives/C02GH833686/p1645710947856339) in order to get familiar with the specifics.
In short `React hooks and refs` are being used to detect and notify when a mouse is up or down. Depending on their states the component gets updated - calling the other functions, new CSS classes etc.

**Review by danjm — DISMISSED** [review]
This looks good to me

**Comment by danjm** [comment]
@mcmire Can you do another review on this?

File: ui/components/app/hold-to-reveal-button/hold-to-reveal-button.js (hunk 4 above)
Comment (mcmire) [reply ui/components/app/hold-to-reveal-button/hold-to-reveal-button.js:35]: Okay, gotcha. It looks like CSS animations power the callbacks that are taking place in the component. That's the piece I was missing :)

File: ui/components/app/hold-to-reveal-button/hold-to-reveal-button.test.js [hunk 5]
```diff
@@ -0,0 +1,35 @@
+import React from 'react';
+import { render, fireEvent } from '@testing-library/react';
+import HoldToRevealButton from './hold-to-reveal-button';
+
+describe('HoldToRevealButton', () => {
+  const onLongPressed = jest.fn();
```
Comment (mcmire) [inline ui/components/app/hold-to-reveal-button/hold-to-reveal-button.test.js:6]: Mind putting this in a `beforeEach`? That way it gets reset each test and doesn't get shared between tests.

File: ui/components/app/hold-to-reveal-button/hold-to-reveal-button.test.js [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+    const { getByText } = render(
+      <HoldToRevealButton
+        buttonText="Hold to reveal SRP"
+        onLongPressed={onLongPressed}
+      />,
+    );
+    expect(getByText('Hold to reveal SRP')).toBeInTheDocument();
+  });
+
+  it('should render a button when mouse is down and up', () => {
```
Comment (mcmire) [inline ui/components/app/hold-to-reveal-button/hold-to-reveal-button.test.js:18]: Any other tests that need to be written here? What happens if you "long press" for the duration of the animation? What happens if you "long press" and lift off before the animation concludes? etc.

**Review by mcmire — COMMENTED** [review]
Have a few more comments on this. If we're dying to get this merged we can revisit the tests later I suppose but given how much logic is going on here it seems like we might want to add a bit more.

File: ui/components/app/hold-to-reveal-button/hold-to-reveal-button.test.js (hunk 5 above)
Comment (filipsekulic) [reply ui/components/app/hold-to-reveal-button/hold-to-reveal-button.test.js]: Fixed in: 270557aff1b25a0ce0135093286afd8739462c8d

File: ui/components/app/hold-to-reveal-button/hold-to-reveal-button.test.js (hunk 6 above)
Comment (filipsekulic) [reply ui/components/app/hold-to-reveal-button/hold-to-reveal-button.test.js:23]: Fixed in: 270557aff1b25a0ce0135093286afd8739462c8d

**Review by mcmire — APPROVED** [review]
I would appreciate a review from an extension team member because it's possible there's something I missed but this looks good to me.

**Review by danjm — APPROVED** [review]
LGTM

**Pull request merged**